### PR TITLE
UnitControl: Update unit dropdown design

### PR DESCRIPTION
**Pull request by mirka**
Part of #41973

## What?

Updates the unit dropdown to the new designs for the 40px size variant.

## Why?

To get our components closer to the mockups.

## How?

Add conditional styles for the dropdown depending on the size variant.

## Testing Instructions

1. `npm run storybook:dev`
2. See the stories for UnitControl.

### Known issues

- #42650

## Screenshots or screencast 

https://user-images.githubusercontent.com/555336/181043192-35466198-5e32-47b7-970a-ab559d47a1f6.mp4

### When the unit isn't a dropdown (i.e. single unit case)

Dropdown text has no accent color when the dropdown is inactive

### CODE REVIEW

**Comment by mirka** [comment]
@pablohoneyhoney @jameskoster  Hi 👋 Can we get some guidance on the interaction details (touch target boundary, hover/focus styles) of this unit dropdown? For reference, these are the hover/focus styles we had for the previous design:

Hover style Focus style

Similarly, we'll soon need guidance for the new step button designs in the NumberControl as well. (Not in this PR though)

image

**Comment by jameskoster** [comment]
Assuming the following anatomy:

Screenshot 2022-06-28 at 12 11 49

Where state styling is informed by existing components (`InputControl` and `Button`), we'd end up with the following:

Screenshot 2022-06-28 at 12 14 18

Here's a gif demonstrating the interactions:



To me this seems like a reasonable place to start that avoids changing other components (better handled separately). But I'll defer to Pablo on how to proceed.

File: packages/components/src/unit-control/styles/unit-control-styles.ts
Comment (mirka) [inline packages/components/src/unit-control/styles/unit-control-styles.ts]: Similar to #38969, this was incorrectly rendering in Arial (per Chrome user agent style).

File: packages/components/src/unit-control/types.ts
Comment (mirka) [inline packages/components/src/unit-control/types.ts:62]: Moved to a direct import from `InputControlProps` so it doesn't get out of date.

**Comment by jameskoster** [comment]
An update to the concept above:

It looks like I was using a slightly outdated Figma component with the wrong button size. We have two options for the unit button:

1. Use the Tertiary button variant (36px height)
2. Use the Small button size (24px height)

### Tertiary


### Small


Thinking ahead to the Stepper component mentioned above, I think it makes sense to use the Tertiary variant because its dimensions match the Icon button:

Screenshot 2022-06-29 at 11 06 29

The only small detail we'll need to consider is the color difference:

Screenshot 2022-06-29 at 11 08 57

**Comment by mirka** [comment]
@jameskoster Now for the final touches ✨ Here it is with the exact same hover/focus styles as Tertiary Button:

https://user-images.githubusercontent.com/555336/180580953-90c5a677-8182-4f91-a293-17f81d2e15d1.mp4

(Tertiary button has a 1px inset hover shadow, and a 1.5px outset focus shadow.)

It's a bit cramped. I'll note that, at least from a technical library maintenance standpoint, we don't necessarily have to follow the Button component sizes/styles — we can easily make a new component specifically to use inside text fields.

There are a few ways we can tweak the text field button, like:

- Make it slightly smaller than 36px.
- Change the hover/focus styles.

Feel free to write down the styles you want, or push the changes directly to this branch (I can take care of any final refactoring that may be necessary).

---

## Known issues

- The `text-align` bug in Safari (#42650) is more noticeable in this new design 🥲

**Comment by jameskoster** [comment]
Hmm yeah that is a bit tight isn't it. Especially hovering the unit button after focussing the input.

Looking at the 'Remove color' button from global styles (which is somewhat similar in terms of general UI) we can probably try `small` `icon` Buttons:

Screenshot 2022-07-25 at 10 37 26

Sorry for flip-flopping 😓

**Comment by mirka** [comment]
> Sorry for flip-flopping 😓

No problem at all 😄

Any preferences on how to deal with overflows? A 24px square is pretty small and will already overflow with common units like `em`/`rem`.

Overflowing with 'em' unit

We can for example set some kind of `max-width`, and grow to the longest option label within that max.

https://user-images.githubusercontent.com/555336/180882081-8e70c074-1d18-4a9b-9473-5568d1d17ecb.mp4

This means that most UnitControls will have rectangular dropdown buttons, and the proportions of those rectangles will vary depending on the longest available option in that UnitControl. That subtle variance across UnitControls may not be desirable. In that case it might be better to just set all dropdown widths to the maximum width regardless of the option lengths.

**Comment by ciampo** [comment]
> In that case it might be better to just set all dropdown widths to the maximum width regardless of the option lengths.

That would be the best choice, IMO (plus maybe a conservative, safety `max-width` in case the unit label is extremely long?)

**Comment by jameskoster** [comment]
Oh, I think it would be ok for the button to hug the current value, probably with a 24px min-width, and a max-width as @ciampo suggested.

File: packages/components/src/unit-control/stories/index.tsx
Comment (mirka) [inline packages/components/src/unit-control/stories/index.tsx:29]: Removed this explicit config because Storybook will use a `radio` by default.

File: packages/components/src/unit-control/stories/index.tsx
Comment (mirka) [inline packages/components/src/unit-control/stories/index.tsx:57]: Removed these max-width containers in the stories because we need to know whether the width is unconstrained by default, and we need to be able to test `__unstableInputWidth`.

**Comment by mirka** [comment]
Thanks y'all, I've updated the OP with a screen recording of the latest styles. This PR is now ready for final review.

The Safari issue (#42650) is pretty annoying. I don't have the appetite to address it in this PR, but we might eventually need to move to a custom select implementation if we want the centered design in Safari. It's unlikely that the bug will be fixed in Safari itself — that bug report just celebrated its 12th birthday 😇

File: packages/components/src/unit-control/styles/unit-control-styles.ts
Comment (ciampo) [inline packages/components/src/unit-control/styles/unit-control-styles.ts:78]: I still don't have a clear idea of when we should use the `space` utility and when we should just write "raw" px values. What do you think?

File: packages/components/src/unit-control/styles/unit-control-styles.ts
Comment (ciampo) [inline packages/components/src/unit-control/styles/unit-control-styles.ts:84]: Thought: I'd love to start standardizing font-sizes across components and be able to provide a "typography palette" through the various components from the library (e.g `Heading`, `Text`..)

File: packages/components/src/unit-control/styles/unit-control-styles.ts
Comment (ciampo) [inline packages/components/src/unit-control/styles/unit-control-styles.ts:137]: Do we need to test the look of the component in "high contrast mode" in any particular way?

**Review by ciampo — COMMENTED** [review]
Code changes LGTM (I just left a few minor comments) 

https://user-images.githubusercontent.com/1083581/181062442-c98eb172-06fe-4f7d-8e57-c5ad4bd04f19.mp4

I'll leave it up to @jameskoster for the final approval

File: packages/components/src/unit-control/styles/unit-control-styles.ts
Comment (mirka) [reply packages/components/src/unit-control/styles/unit-control-styles.ts:78]: 😂 

I think my current rule of thumb is:

- If it's less than 4px, use the exact value
- If it's size variant value, use the exact value (for readability, and because there's a 30px variant!)
- If it's related to spacing (margin/padding), prefer to use `space`

Assuming the main point of `space` is to encourage people to use multiples of 4px, it shouldn't _really_ matter whether the `space` util is used 🤞

File: packages/components/src/unit-control/styles/unit-control-styles.ts
Comment (mirka) [reply packages/components/src/unit-control/styles/unit-control-styles.ts:137]: As far as I know, there isn't a way to test it outside of a Windows environment. The transparent outline thing should be pretty straightforward though.

File: packages/components/src/unit-control/styles/unit-control-styles.ts
Comment (ciampo) [reply packages/components/src/unit-control/styles/unit-control-styles.ts:78]: Sounds reasonable

**Comment by jameskoster** [comment]
Is it possible to render the button at a width that simply hugs the selected value? At the moment width seems to be based on the longest label in the list of `option`s, which can have some awkward manifestations when there's a mixture of short/long labels:

Screenshot 2022-07-27 at 12 49 36

Something like this would be a bit better:

Screenshot 2022-07-27 at 12 51 35

That would enable us to have a greater max-width value as well which I think could be valuable. 48px is a bit tight.

**Comment by mirka** [comment]
> Is it possible to render the button at a width that simply hugs the selected value?

It will be possible if we move to a [custom select implementation](https://github.com/WordPress/gutenberg/pull/42000#issuecomment-1195642180), which will allow us to fix the Safari issue as well. My recommendation is to merge for now with the current implementation, and move that to a follow-up issue since the task scope will be a bit big. (Unless you feel it's a blocker.)

> That would enable us to have a greater max-width value as well which I think could be valuable. 48px is a bit tight.

Yes I was thinking about this too, and it will take a bit of fiddling to get the logic right. The complexity is in maintaining a reasonable flex width balance between the prefix, number, and suffix elements. For example, in a tight context like the Border control:

A BorderControl field that has a prefix, number field, and unit dropdown all in a 110px width

It might even be that we'll need to allow the consumer to customize the width via props, so it makes sense in different contexts.

**Comment by mirka** [comment]
📌 I'm going to hold off on merging this while I look into #42717 first. The unit dropdown looks fine in Storybook but completely breaks down in WP 😭

**Comment by ciampo** [comment]
> Yes I was thinking about this too, and it will take a bit of fiddling to get the logic right. The complexity is in maintaining a reasonable flex width balance between the prefix, number, and suffix elements.

Yeah, I can totally see that. I would personally try to leverage flexbox and model an elastic behavior

> 📌 I'm going to hold off on merging this while I look into https://github.com/WordPress/gutenberg/issues/42717 first. The unit dropdown looks fine in Storybook but completely breaks down in WP 😭

Good catch

**Comment by jameskoster** [comment]
> Unless you feel it's a blocker.

I don't think so :)

**Comment by mirka** [comment]
Ok, I've fixed the #42717 problem, the unit dropdown now looks the same in isolated Storybook and WP 👍

This is now ready for final review.

**Review by ciampo — APPROVED** [review]
LGTM 🚀 

We can look into refining component in follow-ups:

> > Is it possible to render the button at a width that simply hugs the selected value?
> 
> It will be possible if we move to a [custom select implementation](https://github.com/WordPress/gutenberg/pull/42000#issuecomment-1195642180), which will allow us to fix the Safari issue as well. My recommendation is to merge for now with the current implementation, and move that to a follow-up issue since the task scope will be a bit big. (Unless you feel it's a blocker.)
> 
> > That would enable us to have a greater max-width value as well which I think could be valuable. 48px is a bit tight.
> 
> Yes I was thinking about this too, and it will take a bit of fiddling to get the logic right. The complexity is in maintaining a reasonable flex width balance between the prefix, number, and suffix elements. For example, in a tight context like the Border control:
> 
> A BorderControl field that has a prefix, number field, and unit dropdown all in a 110px width
> 
> It might even be that we'll need to allow the consumer to customize the width via props, so it makes sense in different contexts.

**Comment by mirka** [comment]
Follow-ups moved to #42879

**Pull request merged**
